services/victoriametrics/victoriametrics.go

```diff
@@ -0,0 +1,364 @@
+		return nil
+	}
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	return errors.Errorf("%d: %s", resp.StatusCode, b)
+}
+
+func (svc *VictoriaMetrics) loadBaseConfig() *config.Config {
+	var cfg config.Config
+	buf, err := ioutil.ReadFile(svc.baseConfigPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			svc.l.Errorf("Failed to load base victoriametrics config %s: %s", svc.baseConfigPath, err)
+		}
+
+		return &cfg
+	}
+	if err := yaml.Unmarshal(buf, &cfg); err != nil {
+		svc.l.Errorf("Failed to parse base victoriametrics config %s: %s.", svc.baseConfigPath, err)
+
+		return &config.Config{}
+	}
+
+	return &cfg
+}
+
+// marshalConfig marshals VictoriaMetrics configuration.
+func (svc *VictoriaMetrics) marshalConfig() ([]byte, error) {
+	cfg := svc.loadBaseConfig()
+
+	e := svc.db.InTransaction(func(tx *reform.TX) error {
+		settings, err := models.GetSettings(tx)
+		if err != nil {
+			return err
+		}
+		s := settings.MetricsResolutions
+		if cfg.GlobalConfig.ScrapeInterval == 0 {
+			cfg.GlobalConfig.ScrapeInterval = config.Duration(s.LR)
+		}
+		if cfg.GlobalConfig.ScrapeTimeout == 0 {
+			cfg.GlobalConfig.ScrapeTimeout = prometheus.ScrapeTimeout(s.LR)
+		}
+		cfg.ScrapeConfigs = append(cfg.ScrapeConfigs, scrapeConfigForVictoriaMetrics(s.HR), scrapeConfigForVMAlert(s.HR))
+		prometheus.AddInternalServiceToScrape(cfg, s)
+
+		return prometheus.AddScrapeConfigs(svc.l, cfg, tx.Querier, &s)
+	})
+	if e != nil {
+		return nil, e
+	}
+
+	b, err := yaml.Marshal(cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't marshal VictoriaMetrics configuration file")
+	}
+
+	b = append([]byte("# Managed by pmm-managed. DO NOT EDIT.\n---\n"), b...)
+
+	return b, nil
+}
+
+// configAndReload saves given VictoriaMetrics configuration to file and reloads VictoriaMetrics.
+// If configuration can't be reloaded for some reason, old file is restored, and configuration is reloaded again.
+func (svc *VictoriaMetrics) configAndReload(ctx context.Context, cfg []byte) error {
+	oldCfg, err := ioutil.ReadFile(svc.scrapeConfigPath)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	fi, err := os.Stat(svc.scrapeConfigPath)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// restore old content and reload in case of error
+	var restore bool
+	defer func() {
+		if restore {
+			if err = ioutil.WriteFile(svc.scrapeConfigPath, oldCfg, fi.Mode()); err != nil {
+				svc.l.Error(err)
+			}
+			if err = svc.reload(ctx); err != nil {
+				svc.l.Error(err)
+			}
+		}
+	}()
+
+	// write new content to temporary file, check it
+	f, err := ioutil.TempFile("", "pmm-managed-config-victoriametrics-")
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	if _, err = f.Write(cfg); err != nil {
+		return errors.WithStack(err)
+	}
+	defer func() {
+		_ = f.Close()
+		_ = os.Remove(f.Name())
+	}()
+	args := []string{"check", "config", f.Name()}
+	cmd := exec.CommandContext(ctx, "promtool", args...) //nolint:gosec
```

Comment:
don't you have separate tool to check config?

Reply:
We can check it with vmdb --dry-run, but current configuration is incomplete with our checker. So it will fail each time. Later i can replace it, but  in case of backward-capability, i decided to use promtool.

Reply:
ok